src/cargo/core/compiler/build_runner/compilation_files.rs

```diff
@@ -604,11 +604,20 @@ fn compute_metadata(
 
     METADATA_VERSION.hash(&mut shared_hasher);
 
+    let ws_root = if unit.is_std {
+        // SourceId for stdlib crates is an absolute path inside the sysroot.
+        // Pass the sysroot as workspace root so that we hash a relative path.
+        // This avoids the metadata hash changing depending on where the user installed rustc.
+        &bcx.target_data.get_info(unit.kind).unwrap().sysroot
```

Comment:
Just a note for future self: sysroot is something like
`/home/user/.rustup/toolchains/nightly-aarch64-apple-darwin`

It is not the real workspace root path of the standard library.
While this seems a bit out of sync, the relative path inside sysroot should be fairly deterministic. It's fine now and I don't think we need any hardcoded path logic like [this](https://github.com/rust-lang/cargo/blob/addcc8ca715bc7fe20df66afd6efbf3c77ef43f8/src/cargo/core/compiler/standard_lib.rs?plain=1#L217).